filters/auth/grantlogout.go

```diff
@@ -0,0 +1,192 @@
+//
+// grantLogout filter
+//
+// Used for revoking access/refresh tokens generated by the oauthGrant() filter.
+//
+
+package auth
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/zalando/skipper/filters"
+)
+
+const (
+	GrantLogoutName           = "grantLogout"
+	revokeTokenKey            = "token"
+	revokeTokenTypeKey        = "token_type_hint"
+	refreshTokenType          = "refresh_token"
+	accessTokenType           = "access_token"
+	unsupportedTokenTypeError = "unsupported_token_type"
+)
+
+type grantLogoutSpec struct {
+	config OAuthConfig
+}
+
+type grantLogoutFilter struct {
+	config OAuthConfig
+}
+
+type revokeErrorResponse struct {
+	Error            string `json:"error"`
+	ErrorDescription string `json:"error_description"`
+}
+
+func (*grantLogoutSpec) Name() string { return GrantLogoutName }
+
+func (s *grantLogoutSpec) CreateFilter([]interface{}) (filters.Filter, error) {
+	return &grantLogoutFilter{
+		config: s.config,
+	}, nil
+}
+
+func (f *grantLogoutFilter) getBasicAuthCredentials() (string, string, error) {
+	clientID := f.config.GetClientID()
+	if clientID == "" {
+		return "", "", errors.New("failed to create token revoke auth header: no client ID")
+	}
+
+	clientSecret := f.config.GetClientSecret()
+	if clientSecret == "" {
+		return "", "", errors.New("failed to create token revoke auth header: no client secret")
+	}
+
+	return clientID, clientSecret, nil
+}
+
+func responseToError(responseData []byte, statusCode int, tokenType string) error {
+	var errorResponse revokeErrorResponse
+	err := json.Unmarshal(responseData, &errorResponse)
+
+	if err != nil {
+		return err
+	}
+
+	if errorResponse.Error == unsupportedTokenTypeError && tokenType == accessTokenType {
+		// Provider does not support revoking access tokens, which can happen according to RFC 7009.
+		// In that case this is not really an error.
+		return nil
+	}
+	return fmt.Errorf(
+		"%s revocation failed: %d %s: %s",
+		tokenType,
+		statusCode,
+		errorResponse.Error,
+		errorResponse.ErrorDescription,
+	)
+}
+
+func (f *grantLogoutFilter) revokeTokenType(tokenType string, token string) error {
+	revokeURL, err := url.Parse(f.config.RevokeTokenURL)
+	if err != nil {
+		return err
+	}
+
+	query := revokeURL.Query()
+	for k, v := range f.config.AuthURLParameters {
+		query.Set(k, v)
+	}
+	revokeURL.RawQuery = query.Encode()
+
+	body := url.Values{}
+	body.Add(revokeTokenKey, token)
+	body.Add(revokeTokenTypeKey, tokenType)
+
+	revokeRequest, err := http.NewRequest(
+		"POST",
+		revokeURL.String(),
+		strings.NewReader(body.Encode()))
+
+	if err != nil {
+		return err
+	}
+
+	clientId, clientSecret, err := f.getBasicAuthCredentials()
+	if err != nil {
+		return err
+	}
+
+	revokeRequest.SetBasicAuth(clientId, clientSecret)
+	revokeRequest.Header.Add("Content-Type", "application/x-www-form-urlencoded")
```

Comment:
Header.Add -> Header.Set

I guess it's empty in this case, but I would not like to have 2 Content-Type headers if there will be some day a default set

Reply:
Ah, good call. Done ✅